src/Validation.hs

```diff
@@ -62,86 +61,202 @@ module Validation
+Success "John"
+>>> validateName ""
+Failure (EmptyName :| [])
+
+Now we can define all checks for the password separately:
 
 >>> :{
-validateCpus :: Int -> Validation [String] Int
-validateCpus cpus
-    | cpus >= 2 = Success cpus
-    | otherwise = Failure ["Not enough CPUs"]
+validateShortPassword :: String -> Validation (NonEmpty FormValidationError) Password
+validateShortPassword password
+    | length password < 8 = Failure (ShortPassword :| [])
+    | otherwise = Success (Password password)
 :}
 
-You can use these functions with the 'Applicative' instance of the 'Validation'
-type to construct a validated @Computer@. You will get either (pun intended) a
-valid @Computer@ or the errors that prevent it from being considered valid.
+>>> :{
+validatePasswordDigit :: String -> Validation (NonEmpty FormValidationError) Password
+validatePasswordDigit password
+    | any isDigit password = Success (Password password)
+    | otherwise = Failure (NoDigitPassword :| [])
+:}
 
-Like so:
+Now we can easily compose those two checks to validate password:
 
 >>> :{
-mkComputer :: Int -> Int -> Validation [String] Computer
-mkComputer ram cpus = Computer
-    <$> validateRam ram
-    <*> validateCpus cpus
+validatePassword :: String -> Validation (NonEmpty FormValidationError) Password
+validatePassword password = validateShortPassword password *> validatePasswordDigit password
 :}
 
-Using @mkComputer@ we get a @Success Computer@ or a list with all possible errors:
+Let's see how it works:
 
->>> mkComputer 16 2
-Success (Computer {computerRam = 16, computerCpus = 2})
+>>> validatePassword "abcd"
+Failure (ShortPassword :| [NoDigitPassword])
+>>> validatePassword "abcd1"
+Failure (ShortPassword :| [])
+>>> validatePassword "abcd12345"
+Success "abcd12345"
 
->>> mkComputer 16 1
-Failure ["Not enough CPUs"]
+After we've implemented validations for all fields, we can compose
+them together:
 
->>> mkComputer 15 2
-Failure ["Not enough RAM"]
+>>> :{
+validateForm :: Form -> Validation (NonEmpty FormValidationError) User
+validateForm Form{..} = User
+    <$> validateName formUserName
+    <*> validatePassword formPassword
+:}
+
+And it works like a charm:
 
->>> mkComputer 15 1
-Failure ["Not enough RAM","Not enough CPUs"]
+>>> validateForm (Form "" "")
+Failure (EmptyName :| [ShortPassword,NoDigitPassword])
+>>> validateForm (Form "John" "abc")
+Failure (ShortPassword :| [NoDigitPassword])
+>>> validateForm (Form "Jonh" "qwertypassword")
+Failure (NoDigitPassword :| [])
+>>> validateForm (Form "Jonh" "qwertypassword123")
+Success (User {userName = "Jonh", userPassword = "qwertypassword123"})
 -}
 
--- | 'Validation' is 'Either' with a 'Left' that is a 'Semigroup'.
+{- | 'Validation' is a polymorphic sum type for storing either all
+validation failures or validation success. Unlike 'Either', which
+returns only the first error, 'Validation' accumulates all errors
+using the 'Semigroup' typeclass.
+
+Usually @'Validation' e a@ is used like:
```

Comment:
Maybe it worth specifying what `e` field is used like (list or non-empty list). At the moment it looks like everyone should only use error of `String` and success of `User` which is not the most common use case.